### ISSUE TITLE
Epoll reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1953]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#1953|
|Type|enhancement|
|Created|2006-07-24 09:15:52Z|

Using the work done here: [sandbox/exarkun/epoll/](sandbox/exarkun/epoll/) , I reached a working reactor (tests suite pass). I attach it here if anyone interested.

We can see [#1930](https://github.com/twisted/twisted/issues/1930) for another alternative.

Attachments:

* [epollreactor2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/866/866421ad478e2391535b79572800f4c5aae68778/01e9cf8949076d1ed716ec741b60f9cabee62b5c.py) (5957 bytes) - added by therve on 2006-07-27 08:41:45Z - Updated version with more docs and use of the C module
* [epoll.c](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/866/866421ad478e2391535b79572800f4c5aae68778/b164d0331f0d0fa0a9bdc4cfb56a88f46b1548b3.c) (4476 bytes) - added by therve on 2006-07-27 09:04:39Z - Epoll C wrapper
* [epollreactor.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/866/866421ad478e2391535b79572800f4c5aae68778/c4a798cb0c3a629de962b98fea93586b1db29ad2.py) (5947 bytes) - added by therve on 2006-08-03 08:57:37Z - Add more docs
* [_epoll.pyx](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/866/866421ad478e2391535b79572800f4c5aae68778/3e879c86cad02641a5eae896e0cef2a5e9fdb96a.pyx) (4549 bytes) - added by therve on 2006-08-03 08:57:54Z - 
* [test_epoll.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/866/866421ad478e2391535b79572800f4c5aae68778/101d4683ed174fa4cf4c143832478a15a6a08754.py) (3242 bytes) - added by therve on 2006-08-03 08:58:08Z - 
* [1930_app.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/866/866421ad478e2391535b79572800f4c5aae68778/02b4f68a11584ed951a5dcb72d2ada3ae06945f3.diff) (536 bytes) - added by therve on 2006-08-03 09:13:31Z - 
* [1930_setup.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/866/866421ad478e2391535b79572800f4c5aae68778/7ed03c0d569f5afcd98850e92805c6bc753f43f5.diff) (669 bytes) - added by therve on 2006-08-03 09:13:48Z - 
* [reactor_1953.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/866/866421ad478e2391535b79572800f4c5aae68778/38c930c59e7f4d1e382f8f65ca355436f4902479.diff) (2311 bytes) - added by therve on 2006-08-07 10:36:56Z - Diff agains branches/epollreactor-1953 to make the test suite pass
* [comment1_1953.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/866/866421ad478e2391535b79572800f4c5aae68778/b7027ca0a42de4d8ca54142b95c2746b4cdc457d.diff) (6551 bytes) - added by therve on 2006-09-30 07:36:08Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1953 1953
type__enhancement enhancement
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1153732552000000 1153732552000000
changetime__1161631161000000 1161631161000000
version__None None
owner__ 
cc__therve cc__spiv cc__jknight cc__exarkun cc__ralphm cc__PenguinOfDoom
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

This deserves a review tag, even though it's not a proper diff, and clearly needs a bit of polish (allowing --reactor args of twistd, trial to use it, for instance, and updating the docs).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

It would be better for epollreactor to be edge triggered.  The code in my sandbox was level triggered, and it looks like the attached reactor is as well.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

It does need polish, and I'm ready to make it if it encounters some interest.

The attached reactor is indeed Level Triggered, it's just some modification away from the sandbox one, to make the tests suite pass. Creating an Edge Triggered reactor would be much more work: it's really different from select/poll, and I don't know if it performs much better. Some arguments ?

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Yays.

I think we should add a level triggered epoll reactor before going off and trying to make it edge triggered. (Which probably performs better but will likely be a *lot* of work to make work in twisted).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

Also, I wanted to know if the pyrex extension was a requirement. I rewrote a wrapper in plain C which seems to be a lot simpler (even if it may introduce other problems).

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

If it _doesn't_ introduce other problems that's fine with me...pyrex introduces many problems of its own. But it'll have to be carefully checked over to make sure it does all its refcounting/etc right.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

The epoll in my sandbox had unit tests for the low-level epoll wrapper.  Does this C wrapper present the same API (and so is the existing test suite usable to test it)?

In order of preference, I think I'd rather have a wrapper:

  * based on rctypes
  * based on ctypes
  * written in pyrex, perhaps with the resulting C file including svn
  * wrapped directly with C

Modulo pyrex bugs (which may be significant and disqualify this argument), pyrex modules are much easier to write and maintain correctly than C wrappers.

And rctypes may be too raw to actually use :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

I guess the wrapper is close enough to adapt the tests. I'm ready to do it if necessary.

I thought a C wrapper might be better because of performance: my module is around 170 lines against 1000 for the pyrex compiled version. The epoll API is simple enough for making a thin wrapper, but I agree that a pyrex version is easier to maintain.

I'll try to have some performance numbers for C version against pyrex version, but, again, it's just an alternative, I'm sure the pyrex version works very well.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

I wouldn't worry about performance much here right now.  It seems very unlikely to me that the bottleneck in any application is going to be in the epoll bindings.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

You're right, there doesn't seem to be a performance difference between the two version. The choice is then more a maintenance question, so I guess the Pyrex version is more suitable for that.

So, anything else needed to push for its review?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

Mostly just the tests, I think.  There are a few other changes to make, likely: documentation, updating the reactors dictionary, etc, but that is all pretty simple.  Re-add review keyword when you're ready.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

Attach updated version with more docs, and the tests previously done. Also add diffs to setup.py and app.py to build the module and add the epoll reactor to twistd/trial. 

To build I only checked the platform for linux2, but I guess epoll is available on other platform (bsd? darwin?).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

Checked into svn in the epollreactor-1953 branch.  The test suite has a large number of failures, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve commented|
|-|-|

Replying to [exarkun](#note_14):

There are strange behaviours during the test suite. It seems each test run alone but not in a whole. I'll try to find a reproductible behaviour, but I guess it comes from some exceptions raised and bad managed in the reactor.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

Hmm, I took some dubious looking error handling out of the branch and it broke a handful of tests.  I'll take a look at what's going on there soon, but I haven't had a chance yet.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve commented|
|-|-|

Replying to [exarkun](#note_18):
> Hmm, I took some dubious looking error handling out of the branch and it broke a handful of tests.  I'll take a look at what's going on there soon, but I haven't had a chance yet.

No problem. On my dev box the branch generates the same errors with epoll and with poll, so I thought it was spurious problems. I'll also have a look.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve commented|
|-|-|

Replying to [exarkun](#note_18):

OK I have 5 failing tests at home; they come from the epoll_ctl call in _remove. The tests are the following: 
 * twisted.test.test_tcp.LoopbackTestCase.testFailing
 * twisted.test.test_tcp.ConnectorTestCase.testReconnect
 * twisted.test.test_pb.NewCredTestCase.testClientConnectionLost
 * twisted.test.test_ftp.FTPServerPortDataConnectionTestCase.testPORTCannotConnect
 * twisted.test.test_application.TestInternet2.testConnectionGettingRefused

All comes from a problem in t.i.tcp.Client: it seems the socket is closed before being removed from the reactor. Moving the close after seems to do the trick for epoll, but I haven't tested more.

```
Index: twisted/internet/tcp.py
===================================================================
--- twisted/internet/tcp.py     (revision 18011)
+++ twisted/internet/tcp.py     (working copy)
@@ -457,6 +457,13 @@
             not hasattr(self, "connector")):
             return

+        self.connector.connectionFailed(failure.Failure(err))
+        if hasattr(self, "reactor"):
+            # this doesn't happen if we failed in __init__
+            self.stopReading()
+            self.stopWriting()
+            del self.connector
+
         try:
             self._closeSocket()
         except AttributeError:
@@ -464,13 +471,6 @@
         else:
             del self.socket, self.fileno

-        self.connector.connectionFailed(failure.Failure(err))
-        if hasattr(self, "reactor"):
-            # this doesn't happen if we failed in __init__
-            self.stopReading()
-            self.stopWriting()
-            del self.connector
-
     def createInternetSocket(self):
         """(internal) Create a non-blocking socket using
         self.addressFamily, self.socketType.
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_24"></a>|@jml commented|
|-|-|

So, basically, I'm not competent to review this branch: I've never done anything serious with pyrex or raw sockets, and it's too late in the day to puzzle things out. Here's what I got though.

Matters of form:
 * Add docstring for `t.i.tcp.BaseClient.failIfNotConnected`
 * 80 cols max in twisted/test/test_epoll.py
 * 80 cols max in twisted/internet/epollreactor.py
 * It's not clear why `reads`, `writes` etc are global and passed as default args. Please add a comment explaining why to epollreactor.py

Hunches:
 * The huge timeouts in test_epoll.py make me a little suspicious. 
 * test_controlAndWait looks pretty big, is that _really_ a unit test? It looks like five.

Questions:
 * Is "sys.platform == 'linux2'" an adequate test? I have no idea.
 * Do the tests pass using epollreactor?
 * Should there be an explicit test (presumably in test_tcp.py) for the change to tcp.py?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_25"></a>|@jml commented|
|-|-|

Oops. Forgot:

 * Please update the trial manpage with the additional reactor
 * Well documented branch, thanks :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_26"></a>|@therve commented|
|-|-|

Replying to [jml](#note_24):

>  Add docstring for `t.i.tcp.BaseClient.failIfNotConnected`

Done.

>  80 cols max in twisted/test/test_epoll.py

Done.

>  80 cols max in twisted/internet/epollreactor.py

Done.

>  It's not clear why `reads`, `writes` etc are global and passed as default args. Please add a comment explaining why to epollreactor.py
> 

It comes from pollreactor. I think it's mainly for performance purpose (bind local variables). I'll add a comment.

>  The huge timeouts in test_epoll.py make me a little suspicious. 

It's miliseconds ? 1s doesn't seem too much.

>  test_controlAndWait looks pretty big, is that _really_ a unit test? It looks like five.

Most of the work is to setup the sockets, but I guess it could be split in 2.

>  Is "sys.platform == 'linux2'" an adequate test? I have no idea.

No, it's not, because I think epoll is at least available on OS X. I added this test because it worked for me, but it needed a review by someone more experimented.

>  Do the tests pass using epollreactor?

Yes they do, once applied the patch on tcp.py.

>  Should there be an explicit test (presumably in test_tcp.py) for the change to tcp.py?

That'd be better I guess. I think other reactors should raise an error at the same place (in removeReader/removeWriter), but they don't seem to :).

>  Please update the trial manpage with the additional reactor

Done.

>  Well documented branch, thanks :)

I did my best :). I'll attach the patch when I'll get access to a working svn env.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph commented|
|-|-|

I'm taking the "review" keyword off until that patch is attached.  Please re-apply the keyword when you attach the patch...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_28"></a>|@therve commented|
|-|-|

Here's my patch. Still to do:
 * Find the good test in topfiles/setup.py to choose to build _epoll.
 * Test for failIfNotConnected if appropriate/necessary.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

Applied the patch to the branch.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_30"></a>|@ralphm commented|
|-|-|

So, if there is still a todo, is this branch up for review?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_31"></a>|@therve commented|
|-|-|

I set the review tag hoping someone will find an answer to both questions. I guess I'll try to answer myself.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun set owner to @exarkun|
|-|-|

I don't think the failIfNotConnected change needs explicit tests.

I changed setup.py to check for sys/epoll.h.

Ready for review, I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_34"></a>|PenguinOfDoom set owner to @exarkun|
|-|-|

The "except IOError;" clause in doPoll should probably re-raise, just in case.
setup.py still checks for linux2, and not sys/epoll.h
Other than that, it looks fine.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_35"></a>|PenguinOfDoom commented|
|-|-|

It should re-raise unless it's a EINTR, same as pollreactor.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun removed owner|
|-|-|

Changes made, please re-review.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_38"></a>|PenguinOfDoom set owner to @exarkun|
|-|-|

Looks fine! Fire the commit cannons.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun set status to `closed`|
|-|-|

(In [18461]) Merge epollreactor-1953

Author: therve, exarkun
Reviewer: PenguinOfDoom
Fixes [#1953](https://github.com/twisted/twisted/issues/1953)

Add a level-triggered epoll-based reactor.

Also re-order the removal and the shutdown of sockets from the reactor
in the tcp code shared between reactors in order to avoid possibly
passing a no-longer valid file descriptor to a control API such as
epoll_ctl() or kevent().

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun commented|
|-|-|

(In [18464]) Merge epollreactor-1953

Author: therve, exarkun
Reviewer: PenguinOfDoom
Fixes [#1953](https://github.com/twisted/twisted/issues/1953)

Add a level-triggered epoll-based reactor.

Also re-order the removal and the shutdown of sockets from the reactor
in the tcp code shared between reactors in order to avoid possibly
passing a no-longer valid file descriptor to a control API such as
epoll_ctl() or kevent().

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_41"></a>|@exarkun commented|
|-|-|

(In [18465]) Revert r18464 - test suite regression

Refs [#1953](https://github.com/twisted/twisted/issues/1953)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_42"></a>|@exarkun set status to `reopened`|
|-|-|

For those watching at home, the reasons this ticket should have failed review:

 * tests failed on OS X
 * tests failed on Win32
 * tests failed on Python 2.5

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_43"></a>|@jyknight commented|
|-|-|

I wonder if perhaps it'd be a good idea to have buildbot run tests for commits on all branches. How it is now, one has to press 12 build buttons to test a branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_44"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Yes.  That'd be a great buildbot feature.

Some changes committed to the branch:

  * Skip the tests if _epoll cannot be imported
  * Skip the tests if _epoll.epoll(n) fails with ENOSYS
  * Generate the .c using Pyrex 0.9.4.1 from the lxml svn repository, since it seems to have been updated to work with Python 2.5.

Please re-review. :(

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_45"></a>|@jerub set owner to @exarkun|
|-|-|

Following an interview with exarkun, exarkun is going to add some comments to the code based on some questions I asked, and i am running some tests ( http://twistedmatrix.com/buildbot/reactors/builds/1545 ).

If the tests pass, and the comments are committed, I'm prepared to accept this branch for merging. I'm removing the review flag pending either merging, or build failure.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_46"></a>|@exarkun set status to `closed`|
|-|-|

(In [18475]) Merge epollreactor-1953

Author: therve, exarkun 
Reviewer: PenguinOfDoom, Jerub, jml
Fixes [#1953](https://github.com/twisted/twisted/issues/1953)

Add a level-triggered epoll-based reactor.

Also re-order the removal and the shutdown of sockets from the reactor
in the tcp code shared between reactors in order to avoid possibly passing
a no-longer valid file descriptor to a control API such as epoll_ctl() or
kevent().

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_49"></a>|@therve commented|
|-|-|

It seems there was multiple attempts to commit. What's the last blocking problems ?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_50"></a>|@exarkun commented|
|-|-|

The reason for the last revert was the code for skipping tests for the functionality on platforms which lack epoll was broken.  It may be fixed in the branch now.  Just waiting for more reviewer time.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_51"></a>|@ralphm set owner to @exarkun|
|-|-|

I visually and mechanically tested if the last change by exarkun makes sure the tests are skipped when _epoll is missing. Also made sure the epoll tests pass when _epoll _is_ present.

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_52"></a>|@exarkun set status to `closed`|
|-|-|

(In [18507]) Merge epollreactor-1953-2

Author: exarkun, therve
Reviewer: jml, jerub, penguinofdoom, ralphm
Fixes [#1953](https://github.com/twisted/twisted/issues/1953)

Add a level-triggered epoll-based reactor.

Also re-order the removal and the shutdown of sockets from the reactor in the
tcp code shared between reactors in order to avoid possibly passing a no-longer
valid file descriptor to a control API such as epoll_ctl() or kevent().

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
